packages/react-core/src/filters/viewportFeaturesBinary.js

```diff
@@ -0,0 +1,169 @@
+import bboxPolygon from '@turf/bbox-polygon';
+import intersects from '@turf/boolean-intersects';
+import { prepareViewport, isTileFullyVisible } from './viewportFeatures';
+
+const GEOMETRY_TYPES = Object.freeze({
+  Point: 0,
+  LineString: 1,
+  Polygon: 2
+});
+
+function addIntersectedFeaturesInTile({ map, data, viewportIntersection, type, uniqueIdProperty }) {
+  const indices = getIndices(data);
+  const { positions } = data;
+
+  for (let i = 0; i < indices.length - 1; i++) {
+    const startIndex = indices[i];
+    const endIndex = indices[i + 1];
+    const uniquePropertyValue = getUniquePropertyValue(data, startIndex, uniqueIdProperty, map);
+
+    if (uniquePropertyValue && !map.has(uniquePropertyValue)) {
+      const ringCoordinates = getRingCoordinatesFor(startIndex, endIndex, positions);
+      if (intersects(getFeatureByType(ringCoordinates, type), viewportIntersection)) {
+        map.set(uniquePropertyValue, parseProperties(data, startIndex));
+      }
+    }
+  }
+}
+
+function getIndices(data) {
+  const indices = data.primitivePolygonIndices || data.pathIndices || data.pointIndices;
+  return indices.value;
+}
+
+function getFeatureId(data, startIndex) {
+  return data.featureIds.value[startIndex];
+}
+
+function getPropertiesFromTile(data, startIndex) {
+  const featureId = getFeatureId(data, startIndex);
+  const { properties, numericProps } = data;
+  const result = {
+    properties: properties[featureId],
+    numericProps: {}
+  };
+
+  for (const key in numericProps) {
+    result.numericProps[key] = numericProps[key].value[startIndex];
+  }
+
+  return result;
+}
+
+function parseProperties(data, startIndex) {
+  const { properties, numericProps } = getPropertiesFromTile(data, startIndex);
+  return Object.assign({}, properties, numericProps);
+}
+
+function getUniquePropertyValue (data, startIndex, uniqueIdProperty, map) {
+  const { properties, numericProps } = getPropertiesFromTile(data, startIndex);
+  if (uniqueIdProperty) {
+    return numericProps[uniqueIdProperty] || properties[uniqueIdProperty];
+  }
+  return numericProps['cartodb_id'] || numericProps['geoid'] || properties['cartodb_id'] || properties['geoid'] || (map.size + 1);
+}
+
+function getFeatureByType(coordinates, type) {
+  switch (type) {
```

Comment:
Will this manage multi-geometry correctly without further TYPEs?

Reply:
Yes because in binary format multi-geometry are stored as single-geometry with the same `featureId`